controllers/gateway_controller.go

```diff
@@ -36,18 +39,35 @@ const gatewayServiceLabel = "konghq.com/owned-by-gateway"
 type GatewayReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
+	Log    logr.Logger
 }
 
 func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&gatewayv1beta1.Gateway{}).
+		For(&gatewayv1beta1.Gateway{},
+			builder.WithPredicates(predicate.NewPredicateFuncs(r.gatewayHasMatchingGatewayClass)),
+		).
 		Watches(
 			&source.Kind{Type: &corev1.Service{}},
 			handler.EnqueueRequestsFromMapFunc(mapServiceToGateway),
 		).
 		Complete(r)
 }
 
+func (r *GatewayReconciler) gatewayHasMatchingGatewayClass(obj client.Object) bool {
+	gateway, ok := obj.(*gatewayv1beta1.Gateway)
+	if !ok {
+		r.Log.Error(fmt.Errorf("unexpected object type in gateway watch predicates"), "expected", "*gatewayv1beta1.Gateway", "found", reflect.TypeOf(obj))
+		return false
+	}
+	gatewayClass := &gatewayv1beta1.GatewayClass{}
+	if err := r.Client.Get(context.Background(), client.ObjectKey{Name: string(gateway.Spec.GatewayClassName)}, gatewayClass); err != nil {
+		r.Log.Error(err, "could not retrieve gatewayclass", "gatewayclass", gateway.Spec.GatewayClassName)
+		return false
```

Comment:
Strictly speaking we don't know anything about the error we received at this point. It could be that the object was missing (e.g. `404`) or if the cluster is having a bad day it could actually be a server error (e.g. `5XX`). As such we should try and explicitly check if the error is a `NotFound`:

```suggestion
		if errors.IsNotFound(err) {
				r.Log.Error(err, "gatewayclass not found", "gatewayclass", gateway.Spec.GatewayClassName)
				return false
		}
		return false
```

However if the error is something other than `NotFound` we're in the territory of temporarily failures in the Kubernetes API and such, so returning false for those could result in missing an enqueuing of a valid object while the API server is otherwise having issues, and thus reconciliation would ultimately go stale for that object until it gets another update.

Since the Gateway controller will also do this `GatewayClass` check this predicate is just an optimization for busy clusters, so it's safe and preferable to just enqueue the object anyway when we're not really sure what's going on:

```suggestion
		if errors.IsNotFound(err) {
				r.Log.Error(err, "gatewayclass not found", "gatewayclass", gateway.Spec.GatewayClassName)
				return false
		}
		return true # some other problem retrieving the object, enqueue anyway to avoid dropping
```

As the comment above indicates, I recommend that we return true as a failsafe to avoid "losing track" of objects during API failure scenarios.

Reply:
thanks for the detailed explanation @shaneutt